### PR TITLE
feat: Replace socket.io-client

### LINKED ISSUE #90: What is preventing Event Hub from working in Node?
Currently, according to the documentation, event hub doesn't work with Node.

From http://ftrack-javascript-api.rtd.ftrack.com/en/stable/handling_events.html:
> The event hub in the JavaScript API has some minor differences and lacks some of the features available in the [python counterpart](http://ftrack-python-api.rtd.ftrack.com/en/stable/handling_events.html#handling-events).

However, from my point of view, it looks like the `socket.io` client should work in Node out of the box: https://stackoverflow.com/questions/10703513/node-js-client-for-a-socket-io-server

Can you elaborate on what exactly isn't working for Node?

Because in that case, we could be interested in fixing it and contributing a pull request.

### PR DESCRIPTION
**Pull request by gismya**
- [x] I have added automatic tests where applicable
- [x] The PR title is suitable as a release note
- [x] The PR contains a description of what has been changed
- [x] The description contains manual test instructions

## Changes

Removes the old socket.io client and writes a new implementation using the same communication protocol. Re-implements the parts that we were using, and scraps the rest. None of the removed parts were documented, but it was parts that might have been found by people implementing the API, so it will be released as a part of the V2 release with other breaking changes.

Resolves #90 

## Test

Make sure the code looks good and the test cases seem to cover everything we need. Functionality can be tested in the application by yarn linking.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,9 +54,11 @@
   },
   "homepage": "http://ftrack.com",
   "dependencies": {
+    "isomorphic-ws": "^5.0.0",
     "loglevel": "^1.8.1",
     "moment": "^2.29.4",
-    "uuid": "^9.0.0"
+    "uuid": "^9.0.0",
+    "ws": "^8.13.0"
```
Comment (jimmycallin) [inline package.json:61]: i think we want to add ws as a peerDependency to not require it being installed when used for browser purposes

File: source/simple_socketio.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    this.handlers[eventName].push(eventCallback);
+  }
+  // Emit an event with the given eventName and eventData
+  public emit(eventName: string, eventData: Event["_data"]): void {
+    const payload = {
+      name: eventName,
+      args: [eventData],
+    };
+    const dataString = eventData ? `:::${JSON.stringify(payload)}` : "";
+    this.ws?.send(`${PACKET_TYPES.event}${dataString}`);
```
Comment (jimmycallin) [inline source/simple_socketio.ts:174]: it feels a bit wrong to do nothing if the ws isn't initialized. what did the previous client do?

Comment (gismya) [reply source/simple_socketio.ts:174]: Don't think it was handled, the event_hub code does unsent event queueing. But I can investigate

File: package.json (hunk 1 above)
Comment (gismya) [reply package.json:63]: Yes. Done

File: source/simple_socketio.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    if (this.heartbeatInterval) {
+      clearInterval(this.heartbeatInterval);
+      this.heartbeatInterval = undefined;
+    }
+  }
+
+  public isConnected(): boolean {
+    return this.ws?.readyState === WebSocket.OPEN;
+  }
+  // Reconnect methods, to reconnect to the server if the connection is lost using the same session ID
```
[the remark this answers is not in the document]
Comment (gismya) [reply source/simple_socketio.ts]: Done

File: source/simple_socketio.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  private async initializeWebSocket(): Promise<void> {
+    const sessionId = this.sessionId ?? (await this.fetchSessionId());
+    const urlWithQueryAndSession = `${this.wsUrl}/socket.io/1/websocket/${sessionId}?${this.query}`;
+    this.ws = new WebSocket(urlWithQueryAndSession);
+    // Set transport property as a public alias of the websocket
+    this.socket.transport = this.ws;
+    this.addInitialEventListeners();
+  }
+
+  private addInitialEventListeners(): void {
```
[the remark this answers is not in the document]
Comment (gismya) [reply source/simple_socketio.ts]: Done

**Comment by gismya** [comment]
Looks like our tests don't like ws being a peer dependency. Any way to work around that @jimmycallin?

**Comment by jimmycallin** [comment]
> Looks like our tests don't like ws being a peer dependency. Any way to work around that @jimmycallin?

add it as a devDependency as well!

**Comment by gismya** [comment]
> > Looks like our tests don't like ws being a peer dependency. Any way to work around that @jimmycallin?
> 
> add it as a devDependency as well!

Done.

Worth noting is that the `ws` dependency just being there have no impact on the compiled size - it doesn't get automatically added to the final bundle even when it's in the "normal" dependencies. But this is neater.

File: package.json [hunk 5]
```diff
@@ -54,10 +54,14 @@
...
   "homepage": "http://ftrack.com",
   "dependencies": {
+    "isomorphic-ws": "^5.0.0",
     "loglevel": "^1.8.1",
     "moment": "^2.29.4",
     "uuid": "^9.0.0"
   },
+  "peerDependencies": {
+    "ws": "^8.13.0"
+  },
```
Comment (jimmycallin) [inline package.json:64]: add [peerDependenciesMeta](https://yarnpkg.com/configuration/manifest#peerDependenciesMeta) config to tell that ws is optional as well to not output warnings when installing

File: source/simple_socketio.ts (hunk 2 above)
Comment (gismya) [reply source/simple_socketio.ts]: I've added a simple queue

**Comment by gismya** [comment]
@jimmycallin - didn't really expect a full code review yet. Greatly appreciated. If you feel I should split it into parts for easier reviewing, just say the word.

**Comment by gismya** [comment]
@jimmycallin - something I would appreciate help with is if I could do any improvements to the WebSocket-mocking for the tests. Tried adding `mock-socket`, but it didn't play nice with `msw`.

**Comment by ffMathy** [comment]
We love seeing the progress on this! Truly awesome stuff! We so hope it'll make it into production!

**Comment by gismya** [comment]
I'd say it's looking increasingly more likely, now that I've started writing a test suite and at least in the browser seem to be at approximate feature parity. I need to polish the reconnect feature a little bit, write a few more tests and verify it actually working as it should in Node. Barring catastrophic stops on the way I'm now looking to get this in.

**Comment by ffMathy** [comment]
That is amazing news! Yaaaay!!

**Comment by gismya** [comment]
Considering this ready for review. Starting some integration tests with it myself now. If this is too large to review well, I can try splitting it into smaller pieces.

**Comment by gismya** [comment]
It's getting close. It's currently a bit too eager with reconnecting and that will need to be fixed before final testing can begin.

**Comment by ffMathy** [comment]
Great to hear! Thanks for the regular updates! Ftrack is an amazing company in terms of this kind of stuff! State of the art! 😍

**Comment by ffMathy** [comment]
Is the too eager connecting really preventing release though? I mean, technically, it wasn't even working before - right?

**Comment by gismya** [comment]
That part did work previously and needs to be fixed before this goes forward. It's an issue that affects the use of the event hub in the ftrack application, and therefore we can't have any major regressions there. It shouldn't be a major issue to fix - I just have to find time for it.  But once that is done I'll try to get a pre-release of 2.0 out.

**Comment by ffMathy** [comment]
That sounds great. Is there anything we can do to help?

**Comment by ffMathy** [comment]
Maybe I can at least help resolve the conflicts, but I think that'd require write access.

**Comment by gismya** [comment]
Didn't find quite the time to spend on this that I had hoped, and my Friday afternoon brain didn't quite get me all the way there, but the reconnect behavior is better now. If you've got time @ffMathy it would be very helpful if you could check the test `reconnect method exponentially increase delay for every attempt` and see if the error for that is in the test or in the code.

**Comment by gismya** [comment]
I think I've fixed the reconnect behavior now, but didn't have time to properly test it. I'll also have to look at the code again later to see if it needs to be cleaned up.

**Comment by ffMathy** [comment]
That's awesome stuff! I'll throw an additional review at it ASAP.

**Comment by ffMathy** [comment]
LGTM! 😍

File: test/event_hub.test.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { EventHub } from "../source/event_hub";
```
Comment (emm4back) [inline test/event_hub.test.js:1]: Missing Copyright text

File: source/event_hub.ts [hunk 7]
```diff
@@ -247,7 +237,7 @@ export class EventHub {
    *
    * If timeout is non-zero, the promise will be rejected if the event is not
    * sent before the timeout is reached. Should be specified as seconds and
-   * will default to 10.
+   * will default to 30.
```
Comment (ffMathy) [inline source/event_hub.ts:240]: Isn't this a breaking change?

**Review by ffMathy — APPROVED** [review]
Only one minor comment.

File: source/event_hub.ts (hunk 7 above)
Comment (gismya) [reply source/event_hub.ts:240]: It was simply my noticing that the comment was incorrect.

Comment (ffMathy) [reply source/event_hub.ts:240]: Ah okay. All good then!

**Review by jimmycallin — APPROVED** [review]

**Comment by ffMathy** [comment]
Loving the progress so far! We're cheering for you in my team. Keep up the good work! ❤️

**Comment by ffMathy** [comment]
Yaaaay!

**Pull request merged**